File: College/Second Year/DSA/Ex4_CircularQueue.py
class Queue:
    def __init__(self, size: int):
        self.front = self.rear = -1
        self.size = size
        self.queue = [None] * self.size

    def isFull(self):
        return (self.rear + 1) % self.size == self.front

    def isEmpty(self):
        return self.front == -1

    def enqueue(self, value):
        if not self.isFull():
            if self.front == -1:
                self.front = 0
            self.rear = (self.rear + 1) % self.size
            self.queue[self.rear] = value
        else:
            print("Queue Overflow")

    def dequeue(self):
        if not self.isEmpty():
            x = self.queue[self.front]
            self.queue[self.front] = None
            if self.front == self.rear:
                self.front = self.rear = -1
            else:
                self.front = (self.front + 1) % self.size
            return x
        else:
            print("Queue Underflow")
            return None

    def display(self):
        print(f"Front: {self.front}, Rear: {self.rear}")
        if self.isEmpty():
            print("Queue is empty")
        else:
            if self.rear >= self.front:
                print("Queue:", self.queue[self.front:self.rear + 1])
            else:
                print("Queue:", self.queue[self.front:self.size] + self.queue[0:self.rear + 1])


instance = Queue(size=8)
while True:
    choice = int(input("\nCircular Queue Menu:\n1. Enqueue\n2. Dequeue\n3. Display\n4. Exit\nEnter your choice: "))
    if choice == 1:
        value = input("Enter the value: ")
        instance.enqueue(value)
        print("Enqueued:", value)
    elif choice == 2:
        x = instance.dequeue()
        if x is not None:
            print("Dequeued:", x)
    elif choice == 3:
        instance.display()
    elif choice == 4:
        print("Back")
        break
    else:
        print("Invalid choice")

File: College/Second Year/DSA/Ex4_CircularQueue.java
import java.util.Scanner;

class CircularQueue {
    private int[] queue;
    private int rear = -1;
    private int front = -1;
    private int size;

    public CircularQueue(int size) {
        this.queue = new int[size];
        this.size = size;
    }

    public boolean isEmpty() {
        return front == -1;
    }

    public boolean isFull() {
        return (rear + 1) % size == front;
    }

    public void enqueue(int data) {
        if (!isFull()) {
            if (front == -1) {
                front = 0;
            }
            rear = (rear + 1) % size;
            queue[rear] = data;
            System.out.printf("%d enqueued\n", data);
        } else {
            System.out.println("Queue is full");
        }
    }

    public void dequeue() {
        if (!isEmpty()) {
            int value = queue[front];
            queue[front] = 0;
            if (front == rear) {
                front = rear = -1;
            } else {
                front = (front + 1) % size;
            }
            System.out.printf("%d dequeued\n", value);
        } else {
            System.out.println("Queue is empty");
        }
    }

    public void display() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
        } else {
            System.out.print("Queue: [ ");
            int i = front;
            while (true) {
                System.out.print(queue[i] + " ");
                if (i == rear) {
                    break;
                }
                i = (i + 1) % size;
            }
            System.out.println("]");
        }
    }
}

public class Ex4_CircularQueue {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the size of the circular queue: ");
        int size = scanner.nextInt();
        CircularQueue cq = new CircularQueue(size);

        while (true) {
            System.out.println("\nMenu:\n1. Enqueue\n2. Dequeue\n3. Display\n4. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter the value to enqueue: ");
                    int data = scanner.nextInt();
                    cq.enqueue(data);
                    break;
                case 2:
                    cq.dequeue();
                    break;
                case 3:
                    cq.display();
                    break;
                case 4:
                    System.out.println("Exiting...");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice. Please try again.");
                    break;
            }
        }
    }
}
